vpr/src/place/initial_placement.cpp

```diff
@@ -51,38 +51,214 @@ static vtr::vector<ClusterBlockId, t_block_score> assign_block_scores();
+
+    bool legal;
+
+    int cx_from = -1;
+    int cy_from = -1;
+
+    int region_index;
+
+    std::vector<Region> regions = pr.get_partition_region();
+
+    if (regions.size() > 1) {
+        region_index = vtr::irand(regions.size() - 1);
+    } else {
+        region_index = 0;
+    }
+
+    Region reg = regions[region_index];
+
+    vtr::Rect<int> rect = reg.get_region_rect();
+
+    int min_cx = grid_to_compressed_approx(compressed_block_grid.compressed_to_grid_x, rect.xmin());
+    int min_cy = grid_to_compressed_approx(compressed_block_grid.compressed_to_grid_y, rect.ymin());
+
+    int max_cx = grid_to_compressed_approx(compressed_block_grid.compressed_to_grid_x, rect.xmax());
+    int max_cy = grid_to_compressed_approx(compressed_block_grid.compressed_to_grid_y, rect.ymax());
+
+    int delta_cx = max_cx - min_cx;
+
+    int cx_to;
+    int cy_to;
+
+    legal = find_compatible_compressed_loc_in_range(block_type, min_cx, max_cx, min_cy, max_cy, delta_cx, cx_from, cy_from, cx_to, cy_to, false);
+    if (!legal) {
+        //No valid position found
+        return false;
+    }
+
+    compressed_grid_to_loc(block_type, cx_to, cy_to, loc);
+
+    auto& grid = g_vpr_ctx.device().grid;
+
+    VTR_ASSERT_MSG(grid[loc.x][loc.y].width_offset == 0, "Should be at block base location");
+    VTR_ASSERT_MSG(grid[loc.x][loc.y].height_offset == 0, "Should be at block base location");
+
+    return legal;
+}
+
+static bool try_all_part_region_locations(ClusterBlockId blk_id, PartitionRegion& pr, t_logical_block_type_ptr block_type, enum e_pad_loc_type pad_loc_type) {
+    const auto& compressed_block_grid = g_vpr_ctx.placement().compressed_block_grids[block_type->index];
+    auto& place_ctx = g_vpr_ctx.mutable_placement();
+
+    std::vector<Region> regions = pr.get_partition_region();
+
+    bool placed = false;
+
+    t_pl_loc to_loc;
+
+    for (unsigned int reg = 0; reg < regions.size() && placed == false; reg++) {
+        vtr::Rect<int> rect = regions[reg].get_region_rect();
+
+        int min_cx = grid_to_compressed_approx(compressed_block_grid.compressed_to_grid_x, rect.xmin());
+        int max_cx = grid_to_compressed_approx(compressed_block_grid.compressed_to_grid_x, rect.xmax());
+
+        for (int cx = min_cx; cx <= max_cx && placed == false; cx++) {
+            auto y_lower_iter = compressed_block_grid.grid[cx].begin();
+            auto y_upper_iter = compressed_block_grid.grid[cx].end();
+
+            int y_range = std::distance(y_lower_iter, y_upper_iter);
+
+            for (int dy = 0; dy < y_range && placed == false; dy++) {
+                int cy = (y_lower_iter + dy)->first;
+
+                to_loc.x = compressed_block_grid.compressed_to_grid_x[cx];
+                to_loc.y = compressed_block_grid.compressed_to_grid_y[cy];
+
+                auto& grid = g_vpr_ctx.device().grid;
+                auto tile_type = grid[to_loc.x][to_loc.y].type;
+
+                if (regions[reg].get_sub_tile() != NO_SUBTILE) {
```

Comment:
check_subtile helper function

Reply:
Doesn't work in the code - doesn't make exhaustive placement functions any shorter and also cannot be called from both macro and non-macro exhaustive placement functions because of the differences between how subtiles are checked between them.